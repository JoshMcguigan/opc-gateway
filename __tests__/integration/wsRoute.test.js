const WebSocket = require('ws');

let ws;

beforeEach(()=>{
    ws = new WebSocket('ws://localhost:3000/api');
});

test('Server accepts websockets connection at /api', (done)=>{
    ws.on('open', function() {
        done();
    });
});

test('return status code 200 reading multiple variables', (done) => {
    let requestData = {
        "readVariableValue": ["ns=1;s=FanSpeed", "ns=1;s=PumpSpeed", "ns=1;s=SomeDate"]
    };
    ws.on('open', function open() {
        ws.send(JSON.stringify(requestData));
    });
    ws.on('message', function(response){
        let responseObject = JSON.parse(response);
        expect(responseObject.statusCode).toBe(200);
        done();
    });
});

test('return status code 200 reading single variable', (done) => {
    let requestData = {
        "readVariableValue": "ns=1;s=FanSpeed"
    };
    ws.on('open', function open() {
        ws.send(JSON.stringify(requestData));
    });
    ws.on('message', function(response){
        let responseObject = JSON.parse(response);
        expect(responseObject.statusCode).toBe(200);
        done();
    });
});

test('return appropriate data when reading single value', (done)=>{
    // "ns=1;s=SomeDate" used here because the value does not change on the test server
    let requestData = {
        "readVariableValue": "ns=1;s=SomeDate"
    };
    ws.on('open', function open() {
        ws.send(JSON.stringify(requestData));
    });
    ws.on('message', function(response){
        let responseObject = JSON.parse(response);
        expect(responseObject).toMatchSnapshot();
        done();
    });
});

test('return appropriate data when reading multiple values', (done)=>{
    // "ns=1;s=SomeDate" used here because the value does not change on the test server
    let requestData = {
        "readVariableValue": ["ns=1;s=SomeDate", "ns=1;s=SomeDate"]
    };
    ws.on('open', function open() {
        ws.send(JSON.stringify(requestData));
    });
    ws.on('message', function(response){
        let responseObject = JSON.parse(response);
        expect(responseObject).toMatchSnapshot();
        done();
    });
});

test('return status code 200 when requesting browse to single folder which exists', (done)=>{
    // "RootFolder" exists on test server
    let requestData = {
        "browse": "RootFolder"
    };
    ws.on('open', function open() {
        ws.send(JSON.stringify(requestData));
    });
    ws.on('message', function(response){
        let responseObject = JSON.parse(response);
        expect(responseObject.statusCode).toBe(200);
        done();
    });
});

test('return status code 200 when requesting browse to multiple folders which exist', (done)=>{
    // "RootFolder" exists on test server
    let requestData = {
        "browse": ["RootFolder", "RootFolder"]
    };
    ws.on('open', function open() {
        ws.send(JSON.stringify(requestData));
    });
    ws.on('message', function(response){
        let responseObject = JSON.parse(response);
        expect(responseObject.statusCode).toBe(200);
        done();
    });
});

test('return status code 502 when requesting browse to folder which does not exist', (done)=>{
    let requestData = {
        "browse": "madeUpFolderName"
    };
    ws.on('open', function open() {
        ws.send(JSON.stringify(requestData));
    });
    ws.on('message', function(response){
        let responseObject = JSON.parse(response);
        expect(responseObject.statusCode).toBe(502);
        done();
    });
});

test('return appropriate data when requesting browse to single folder which exists', (done)=>{
    // "RootFolder" exists on test server
    let requestData = {
        "browse": "RootFolder"
    };
    ws.on('open', function open() {
        ws.send(JSON.stringify(requestData));
    });
    ws.on('message', function(response){
        let responseObject = JSON.parse(response);
        expect(responseObject).toMatchSnapshot();
        done();
    });
});

test('return appropriate data when requesting browse to multiple folders which exist', (done)=>{
    // "RootFolder" exists on test server
    let requestData = {
        "browse": ["RootFolder", "RootFolder"]
    };
    ws.on('open', function open() {
        ws.send(JSON.stringify(requestData));
    });
    ws.on('message', function(response){
        let responseObject = JSON.parse(response);
        expect(responseObject).toMatchSnapshot();
        done();
    });
});

test('return appropriate data when browsing and reading in single request', (done)=>{
    let requestData = {
        "browse": "RootFolder",
        "readVariableValue": "ns=1;s=SomeDate"
    };
    ws.on('open', function open() {
        ws.send(JSON.stringify(requestData));
    });
    ws.on('message', function(response){
        let responseObject = JSON.parse(response);
        expect(responseObject).toMatchSnapshot();
        done();
    });
});
