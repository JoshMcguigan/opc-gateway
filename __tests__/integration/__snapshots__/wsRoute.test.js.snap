// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`return appropriate data when browsing and reading in single request 1`] = `
Object {
  "body": Object {
    "browse": Array [
      Object {
        "references": Array [
          Object {
            "browseName": Object {
              "name": "FolderType",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "FolderType",
            },
            "isForward": true,
            "nodeClass": "ObjectType",
            "nodeId": "ns=0;i=61",
            "referenceTypeId": "ns=0;i=40",
            "typeDefinition": "ns=0;i=0",
          },
          Object {
            "browseName": Object {
              "name": "Objects",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "Objects",
            },
            "isForward": true,
            "nodeClass": "Object",
            "nodeId": "ns=0;i=85",
            "referenceTypeId": "ns=0;i=35",
            "typeDefinition": "ns=0;i=61",
          },
          Object {
            "browseName": Object {
              "name": "Types",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "Types",
            },
            "isForward": true,
            "nodeClass": "Object",
            "nodeId": "ns=0;i=86",
            "referenceTypeId": "ns=0;i=35",
            "typeDefinition": "ns=0;i=61",
          },
          Object {
            "browseName": Object {
              "name": "Views",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "Views",
            },
            "isForward": true,
            "nodeClass": "Object",
            "nodeId": "ns=0;i=87",
            "referenceTypeId": "ns=0;i=35",
            "typeDefinition": "ns=0;i=61",
          },
        ],
        "statusCode": Object {
          "description": "No Error",
          "name": "Good",
          "value": 0,
        },
      },
    ],
    "readVariableValue": Object {
      "serverPicoseconds": 0,
      "sourcePicoseconds": 0,
      "statusCode": Object {
        "description": "No Error",
        "name": "Good",
        "value": 0,
      },
      "value": Object {
        "arrayType": "Scalar",
        "dataType": "DateTime",
        "value": "2016-10-13T08:40:00.000Z",
      },
    },
  },
  "request": "{\\"browse\\":\\"RootFolder\\",\\"readVariableValue\\":\\"ns=1;s=SomeDate\\"}",
  "statusCode": 200,
}
`;

exports[`return appropriate data when reading multiple values 1`] = `
Object {
  "body": Object {
    "readVariableValue": Array [
      Object {
        "serverPicoseconds": 0,
        "sourcePicoseconds": 0,
        "statusCode": Object {
          "description": "No Error",
          "name": "Good",
          "value": 0,
        },
        "value": Object {
          "arrayType": "Scalar",
          "dataType": "DateTime",
          "value": "2016-10-13T08:40:00.000Z",
        },
      },
      Object {
        "serverPicoseconds": 0,
        "sourcePicoseconds": 0,
        "statusCode": Object {
          "description": "No Error",
          "name": "Good",
          "value": 0,
        },
        "value": Object {
          "arrayType": "Scalar",
          "dataType": "DateTime",
          "value": "2016-10-13T08:40:00.000Z",
        },
      },
    ],
  },
  "request": "{\\"readVariableValue\\":[\\"ns=1;s=SomeDate\\",\\"ns=1;s=SomeDate\\"]}",
  "statusCode": 200,
}
`;

exports[`return appropriate data when reading single value 1`] = `
Object {
  "body": Object {
    "readVariableValue": Object {
      "serverPicoseconds": 0,
      "sourcePicoseconds": 0,
      "statusCode": Object {
        "description": "No Error",
        "name": "Good",
        "value": 0,
      },
      "value": Object {
        "arrayType": "Scalar",
        "dataType": "DateTime",
        "value": "2016-10-13T08:40:00.000Z",
      },
    },
  },
  "request": "{\\"readVariableValue\\":\\"ns=1;s=SomeDate\\"}",
  "statusCode": 200,
}
`;

exports[`return appropriate data when requesting browse to multiple folders which exist 1`] = `
Object {
  "body": Object {
    "browse": Array [
      Object {
        "references": Array [
          Object {
            "browseName": Object {
              "name": "FolderType",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "FolderType",
            },
            "isForward": true,
            "nodeClass": "ObjectType",
            "nodeId": "ns=0;i=61",
            "referenceTypeId": "ns=0;i=40",
            "typeDefinition": "ns=0;i=0",
          },
          Object {
            "browseName": Object {
              "name": "Objects",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "Objects",
            },
            "isForward": true,
            "nodeClass": "Object",
            "nodeId": "ns=0;i=85",
            "referenceTypeId": "ns=0;i=35",
            "typeDefinition": "ns=0;i=61",
          },
          Object {
            "browseName": Object {
              "name": "Types",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "Types",
            },
            "isForward": true,
            "nodeClass": "Object",
            "nodeId": "ns=0;i=86",
            "referenceTypeId": "ns=0;i=35",
            "typeDefinition": "ns=0;i=61",
          },
          Object {
            "browseName": Object {
              "name": "Views",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "Views",
            },
            "isForward": true,
            "nodeClass": "Object",
            "nodeId": "ns=0;i=87",
            "referenceTypeId": "ns=0;i=35",
            "typeDefinition": "ns=0;i=61",
          },
        ],
        "statusCode": Object {
          "description": "No Error",
          "name": "Good",
          "value": 0,
        },
      },
      Object {
        "references": Array [
          Object {
            "browseName": Object {
              "name": "FolderType",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "FolderType",
            },
            "isForward": true,
            "nodeClass": "ObjectType",
            "nodeId": "ns=0;i=61",
            "referenceTypeId": "ns=0;i=40",
            "typeDefinition": "ns=0;i=0",
          },
          Object {
            "browseName": Object {
              "name": "Objects",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "Objects",
            },
            "isForward": true,
            "nodeClass": "Object",
            "nodeId": "ns=0;i=85",
            "referenceTypeId": "ns=0;i=35",
            "typeDefinition": "ns=0;i=61",
          },
          Object {
            "browseName": Object {
              "name": "Types",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "Types",
            },
            "isForward": true,
            "nodeClass": "Object",
            "nodeId": "ns=0;i=86",
            "referenceTypeId": "ns=0;i=35",
            "typeDefinition": "ns=0;i=61",
          },
          Object {
            "browseName": Object {
              "name": "Views",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "Views",
            },
            "isForward": true,
            "nodeClass": "Object",
            "nodeId": "ns=0;i=87",
            "referenceTypeId": "ns=0;i=35",
            "typeDefinition": "ns=0;i=61",
          },
        ],
        "statusCode": Object {
          "description": "No Error",
          "name": "Good",
          "value": 0,
        },
      },
    ],
  },
  "request": "{\\"browse\\":[\\"RootFolder\\",\\"RootFolder\\"]}",
  "statusCode": 200,
}
`;

exports[`return appropriate data when requesting browse to single folder which exists 1`] = `
Object {
  "body": Object {
    "browse": Array [
      Object {
        "references": Array [
          Object {
            "browseName": Object {
              "name": "FolderType",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "FolderType",
            },
            "isForward": true,
            "nodeClass": "ObjectType",
            "nodeId": "ns=0;i=61",
            "referenceTypeId": "ns=0;i=40",
            "typeDefinition": "ns=0;i=0",
          },
          Object {
            "browseName": Object {
              "name": "Objects",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "Objects",
            },
            "isForward": true,
            "nodeClass": "Object",
            "nodeId": "ns=0;i=85",
            "referenceTypeId": "ns=0;i=35",
            "typeDefinition": "ns=0;i=61",
          },
          Object {
            "browseName": Object {
              "name": "Types",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "Types",
            },
            "isForward": true,
            "nodeClass": "Object",
            "nodeId": "ns=0;i=86",
            "referenceTypeId": "ns=0;i=35",
            "typeDefinition": "ns=0;i=61",
          },
          Object {
            "browseName": Object {
              "name": "Views",
              "namespaceIndex": 0,
            },
            "displayName": Object {
              "text": "Views",
            },
            "isForward": true,
            "nodeClass": "Object",
            "nodeId": "ns=0;i=87",
            "referenceTypeId": "ns=0;i=35",
            "typeDefinition": "ns=0;i=61",
          },
        ],
        "statusCode": Object {
          "description": "No Error",
          "name": "Good",
          "value": 0,
        },
      },
    ],
  },
  "request": "{\\"browse\\":\\"RootFolder\\"}",
  "statusCode": 200,
}
`;
